Rubberduck.Parsing/Rewriter/RewritingManager.cs

```diff
@@ -0,0 +1,93 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace Rubberduck.Parsing.Rewriter
+{
+    public class RewritingManager : IRewritingManager
+    {
+        private readonly HashSet<IRewriteSession> _activeCodePaneSessions = new HashSet<IRewriteSession>();
+        private readonly HashSet<IRewriteSession> _activeAttributesSessions = new HashSet<IRewriteSession>();
+
+        private readonly IRewriteSessionFactory _sessionFactory;
+
+        private readonly object _invalidationLockObject = new object();
+
+        public RewritingManager(IRewriteSessionFactory sessionFactory)
+        {
+            _sessionFactory = sessionFactory;
+        }
+
+
+        public IRewriteSession CheckOutCodePaneSession()
+        {
+            var newSession = _sessionFactory.CodePaneSession(TryAllowExclusiveRewrite);
+            lock (_invalidationLockObject)
+            {
+                _activeCodePaneSessions.Add(newSession);
+            }
+
+            return newSession;
+        }
+
+        public IRewriteSession CheckOutAttributesSession()
```

Comment:
I wonder if we need to consider doing a `CheckOutSession(CodeKind codeKind)` instead of explicit `CheckOutAttributesSessions` and `CheckOutCodeSession`? 

It also isn't clear to me how one would request for a rewrite of a module that does not exist -- to support refactoring such as Extract Method where we will have a code preview that may change dramatically based on user's selection on the dialog. The code preview would need to use `IStringParser` - does this support rewriting for that, too?

Reply:
I can see the idea with checking out a session based on a `CodeKind`. However, I do not see what we would gain; more precisely, I do not see what else we would ever need a module rewriter for that might conflict with other rewriting.

Please note that an `IModuleRewriter` has a different function than the pure antlr rewriter it wraps. It is responsible for using its `ISourceCodeHandler` to change the correct component on rewrite. The purpose of the `IRewriteSession` is that multiple sources of rewriting are isolated and to allow the `IRewriteManager` to ensure that only rewriters based on still valid code are applied.

For new code, an `IRewriteSession` does not make too much sense in my opinion: there is nothing to coordinate the rewrite with. Accordingly, for a new module, one would simply generate the content, add a new component and paste the content, just like the `ExtractInterfaceRefactoring`. In case one wants to parse the new contents and then rewrite them, one can use an `IStringParser` and then use an ordinary antlr rewriter.

Btw, the old setup also did not support requesting `IModuleRewriter`s for components that are not loaded.